Lib/fontTools/varLib/instancer/__init__.py

```diff
@@ -785,7 +779,7 @@ def _instantiateFeatureVariations(table, fvarAxes, axisLimits):
             record, i, axisLimits, fvarAxes, axisIndexMap
         )
         if shouldKeep:
-            shouldKeep = _limitFeatureVariationRecord(record, axisLimits, axisOrder)
```

Comment:
@behdad I don't understand this change, can you elaborate what is this "small fix" actually fixing? I think `_limitFeatureVariationRecord` should be given the opportunity to influence whether the record should be kept, that is if any condition is not met because outside the desired range. I think right now you're setting ConditionTable to None, but that's supposed to be a list..

Reply:
Humm. You are right that this might be wrong still. My thinking was that an empty condition-set also means "always apply", and should not cause the variation record to be removed.

Reply:
I think now `_instantiateFeatureVariationRecord` drops any record that is completely irrelevant. So I think the change is correct still. Remove this change and see the test failure.

Reply:
in this case, what's the effect of keeping that empty condition-set that always applies if we have already applied the feature variation as the default feature record? We already swap {"uni0041": "uni0061"}, so what's the use of the empty alwats-apply condition that replaces the same {"uni0041": "uni0061"}?

Reply:
In any case, as it currently stands in this PR, `_limitFeatureVariationRecord` has a couple issues:

- a couple comments are stale: "remapped axis index" (it no longer does that, because `_instantiateFeatureVariationRecord` already did before it); "remove entire record" (it doesn't do that either anymore, but that's under discussion in this thread)
- setting `record.ConditionSet.ConditionTable = newConditions` when `newConditions` might be None,  see that `else` branch which does `newConditions = None`; I think if you're going to keep the `ConditionTable`, the latter is supposed to be a list, empty if you like, but not a `None` value. Besides, that very line does not seem covered by the current test, but I also wonder if it can be _ever_ hit at all (I put an `assert 0` and tried a bunch of axisLimits but couldn't trigger it).

Reply:
> * setting `record.ConditionSet.ConditionTable = newConditions` when `newConditions`

I thought that's fine. It is in HarfBuzz land. But might not be in OpenType land, you are right.





> in this case, what's the effect of keeping that empty condition-set that always applies if we have already applied the feature variation as the default feature record?

Humm. You are right, almost. Unless this rules had the effect of stopping further rules after it, in which case to preserve the logic we need to preserve the rule. Doe you understand what I mean?

Reply:
Ah, I checked the code and you are right, the previous logic was right and I'm totally confused here.

Reply:
Well not exactly. I just need to filter out None. Empty list should still be allowed.

Reply:
I did that now.

Reply:
https://github.com/fonttools/fonttools/commit/95a0fb4a0f933fa3bf0bfc8a60fa68fe12c9e6f7

Reply:
> You are right, almost. Unless this rules had the effect of stopping further rules after it, in which case to preserve the logic we need to preserve the rule. Doe you understand what I mean?

no, but I will sleep over it and return with a fresh mind tomorrow

Reply:
Okay I think I got it: If a rule becomes always-applicable (because the new designspace is a subset of its condition-set), then it should be lifted to default, *and all rules after it should be discarded*.

Reply:
Let's make this code in a shape that it's not more broken that the existing code, then I'll rewrite it completely in a followup PR in a separate module for simplicity.

Reply:
hm ok.

> Let's make this code in a shape that it's not more broken that the existing code,

are you going to do that?

Reply:
I think after https://github.com/fonttools/fonttools/commit/95a0fb4a0f933fa3bf0bfc8a60fa68fe12c9e6f7 it should be roughly in that state.